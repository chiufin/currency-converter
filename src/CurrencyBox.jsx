import React, { useState, useEffect } from 'react';
import { getCurrencyList, getRate } from './apis';

const CurrencyType = {
    FROM: 'from',
    TO: 'to'
}

const CurrencyBox = ({from='USD', to='EUR'}) => {
    const [fromCurrency, setFromCurrency] = useState(from);
    const [toCurrency, setToCurrency] = useState(to);
    const [rate, setRate] = useState(1);
    const [list, setList] = useState([]);
    const [amount, setAmount] = useState(1);

    useEffect(() => {
        getCurrencyList()
        .then( list => setList(list) )
        
        getRate(fromCurrency, toCurrency)
        .then( rate => {
            setRate(rate)
        })
    }, [fromCurrency, toCurrency])

    const changeCurrency = ({ target : { value }}, whichCurrency) => {
        switch(whichCurrency){
            case CurrencyType.FROM:
                setFromCurrency(value);
                break;
            case CurrencyType.TO:
                setToCurrency(value);
                break;
            default:
                break;
        }
    }

    const changeAmount =  ({ target : { value }}, whichCurrency ) => {
        if(!isNaN(Number(value))){
            switch(whichCurrency){
                case CurrencyType.FROM:
                    setAmount(value);
                    break;
                case CurrencyType.TO:
                    setAmount(value*(1/rate));
                    break;
                default:
                    break;
            }
        }
    }
  return (
    <div className="CurrencyBox-container">
        <OneBox type={CurrencyType.FROM}
                from={fromCurrency} 
                to={toCurrency} 
                fromFullName={fromCurrency}
                rate={rate}
                list={list}
                amount={amount}
                changeCurrency={changeCurrency}
                changeAmount={changeAmount}/>
        <OneBox type={CurrencyType.TO}
                from={toCurrency} 
                to={fromCurrency} 
                fromFullName={toCurrency}
                rate={(1/rate)}
                list={list}
                amount={amount*rate}
                changeCurrency={changeCurrency}
                changeAmount={changeAmount}/>
    </div>
  );
}

const OneBox = ({type, from, to, fromFullName, rate, list, amount, changeCurrency, changeAmount}) => 
<div className="CurrencyBox">
    <p className="CurrencyBox-label">1 {from} = {rate} {to}</p>
    <div className="CurrencyBox-input">
        <div className="CurrencyBox-input-dropdown-container">
            <select className="CurrencyBox-input-dropdown"
                    value={ fromFullName }
                    onChange={e => changeCurrency( e, type)}>
                {list.map(each =>
                    <option key={each}>{each}</option>
                )}
            </select>
        </div>
        <input 
            className="CurrencyBox-input-text"
            type="text" 
            value={amount}
            onChange={e => changeAmount( e, type)}>
        </input>
    </div>
</div>

export default CurrencyBox;
