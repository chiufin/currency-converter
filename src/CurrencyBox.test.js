import React from 'react';
import { shallow, mount} from 'enzyme';
import CurrencyBox, { OneBox } from './CurrencyBox';

describe("Currency Box" , () => {
  const wrapper = shallow(<CurrencyBox />);
  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('CurrencyBox', () => {
  
    const wrapper = mount(<CurrencyBox/>);
    expect(wrapper).toBeDefined();
    wrapper.find('input[name="from"]').simulate('change', {
      target: {
        value: 100,
        name: "from"
      },
    });
    wrapper.find('input[name="to"]').simulate('change', {
      target: {
        value: 100,
        name: "to"
      },
    });
    wrapper.find('select[name="from"]').simulate('change', {
      target: {
        value: 'EUR',
        name: "from"
      },
    });
    wrapper.find('select[name="to"]').simulate('change', {
      target: {
        value: 'EUR',
        name: "to"
      },
    });
  });
  it('OneBox: renders from currency, to currency', () => {
    const props = {
      from: 'USD',
      to: 'EUR',
      list: ['USD', 'EUR']
    }
    const wrapper = shallow(<OneBox {...props}/>);
    expect(wrapper.find(props.from)).toBeDefined();
    expect(wrapper.find(props.to)).toBeDefined();

  });
})