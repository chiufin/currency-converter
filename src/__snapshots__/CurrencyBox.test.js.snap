// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency Box should match the snapshot 1`] = `"<div class=\\"CurrencyBox-container\\"><div class=\\"CurrencyBox\\"><p class=\\"CurrencyBox-label\\">1 USD = 1 USD</p><div class=\\"CurrencyBox-input\\"><div class=\\"CurrencyBox-input-dropdown-container\\"><select class=\\"CurrencyBox-input-dropdown\\" name=\\"from\\"></select></div><input type=\\"text\\" class=\\"CurrencyBox-input-text\\" name=\\"from\\" value=\\"1\\"/></div></div><div class=\\"CurrencyBox-convert-container\\"><img class=\\"CurrencyBox-convert-icon\\" src=\\"convert_icon.png\\" alt=\\"covert icon\\"/></div><div class=\\"CurrencyBox\\"><p class=\\"CurrencyBox-label\\">1 USD = 1 USD</p><div class=\\"CurrencyBox-input\\"><div class=\\"CurrencyBox-input-dropdown-container\\"><select class=\\"CurrencyBox-input-dropdown\\" name=\\"to\\"></select></div><input type=\\"text\\" class=\\"CurrencyBox-input-text\\" name=\\"to\\" value=\\"1\\"/></div></div></div>"`;
